vrs_anvil/translator.py

```diff
@@ -0,0 +1,134 @@
+        self.task_queue = task_queue
+        self.result_queue = result_queue
+        self.translator = caching_allele_translator_factory(normalize=normalize)
+        self.busy = False
+
+    def run(self):
+        while True:
+            try:
+                prioritized_item = self.task_queue.get()
+                item: VCFItem = prioritized_item.item
+                if item is None:
+                    break  # Signal to exit the thread
+
+                self.busy = True
+                allele = self.translator.translate_from(fmt=item.fmt, var=item.var)
+                _ = item._asdict()
+                _["result"] = allele
+                result = VCFItem(**_)
+                self.result_queue.put(PrioritizedItem(1, result))
+
+                self.busy = False
+                self.task_queue.task_done()
+
+            except Exception as exc:
+                _logger.exception(f"{self.name} error {exc}")
+                self.busy = False
+                break
+
+
+class VCFItem(NamedTuple):
+    """A named tuple to hold the VCF item."""
+    fmt: str
+    """The format of the item's id e.g gnomad - passed to vrs python"""
+    var: str
+    """variant - passed to vrs python"""
+    file_name: str = None
+    """data source"""
+    line_number: int = None
+    """line number in the file"""
+    identifier: str = None
+    """identifier for the item"""
+    result: Any = None
+    """identifier for the item"""
+
+
+@dataclass(order=True)
+class PrioritizedItem:
+    """If the data elements are not comparable, the data can be wrapped in a class that ignores the data item and only compares the priority number"""
+    priority: int
+    item: Any = field(compare=False)
+
+
+class Translator(BaseModel):
+    """A class to run the translation in either threaded or non-threaded fashion."""
+
+    normalize: Optional[bool] = False
+
+    def translate_from(
+        self, generator: Generator[VCFItem, None, None], num_threads: int = 8
+    ) -> Generator[VCFItem, None, None]:
+        if num_threads > 1:
+            return threaded_translator(generator, num_threads, self.normalize)
+        else:
+            return inline_translator(generator, self.normalize)
+
+
+def inline_translator(generator: Generator[VCFItem, None, None], normalize: bool = False) -> Generator[VCFItem, None, None]:
+    """A generator that runs the translation in a non-threaded fashion."""
+    tlr = caching_allele_translator_factory(normalize=normalize)
+    for item in generator:
+        allele = tlr.translate_from(fmt=item.fmt, var=item.var)
+        _ = item._asdict()
+        _["result"] = allele
+        yield VCFItem(**_)
+
+
+def threaded_translator(generator: Generator[VCFItem, None, None], num_worker_threads: int, normalize: bool = False) -> Generator[VCFItem, None, None]:
+    """A generator that runs the translation in a threaded fashion."""
+    task_queue = queue.PriorityQueue(maxsize=num_worker_threads * 2)
```

Comment:
since we can use Python 3.12 in Terra, can we use a regular queue?

Reply:
True, we are using 3.12, however, we may not have control over what version others may use.

Reply:
so if we've made the change to use a priority queue (works on all python versions), why do we need to setup pyenv in Terra again?